In [ ]:
from huggingface_hub import login

# Provide your Hugging Face token
login(token="your token here")


C:\Users\pooji\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


### Model Loading:
The script is loading the Mistral-7B-Instruct-v0.2 model using Hugging Face's transformers library.

* AutoTokenizer is used to load a pre-trained tokenizer
* AutoModelForCausalLM loads the pre-trained language model for text generation.

In [2]:
# Load model directly

from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-Instruct-v0.2")
model = AutoModelForCausalLM.from_pretrained("mistralai/Mistral-7B-Instruct-v0.2")

Loading checkpoint shards: 100%|██████████| 3/3 [01:12<00:00, 24.13s/it]


### Importing all the required libraries

In [3]:
import pandas as pd
import nltk
from nltk.translate.bleu_score import sentence_bleu
from nltk.translate.meteor_score import meteor_score
import torch
import collections
from collections import Counter
import math
from transformers import GPT2LMHeadModel, GPT2Tokenizer
from nltk.util import ngrams
import pronouncing
from nltk.corpus import cmudict
from collections import defaultdict
import re

### Dataset Handling:

The dataset is loaded from a CSV file containing prompts and generated poems. It includes all the poems generated by six models: Llama 3.2 3B Instruct Q8, Mistral 7B Instruct Q4, Gemma 2 2B Q4, QWEN2 7B Instruct Q4, Openchat-3.5 7B, and Code Ninja 7B Q4. 

Below are the prompts used to generate three poems:
* Prompt 1: Write a Shakespearean sonnet about courage set on a battlefield with a determined tone. Use vivid imagery to convey strength and resilience.
* Prompt 2: Write a Shakespearean sonnet about wonder set in outer space with an awe-filled tone. Use vivid imagery to convey mystery and discovery.
* Prompt 3: Write a Shakespearean sonnet about loss set in an empty home with a somber tone. Use vivid imagery to convey grief and reflection.

In [4]:
file_path = "Model_gen_Poems.csv"
data = pd.read_csv(file_path)
print("Dataset Preview:")
print(data.head())
print("\nColumn Names:")
print(data.columns)
print("\nDataset Info:")
data.info()

Dataset Preview:
                                        Prompt/Model  \
0  Write a Shakespearean sonnet \nabout courage s...   
1  Write a Shakespearean sonnet \nabout wonder se...   
2  Write a Shakespearean sonnet \nabout loss set ...   

                            Llama 3.2 3B Instruct Q8  \
0  Fair battlefield, where valorous hearts do lie...   
1  Fairest cosmos, thou dost stretch thy might,\n...   
2  In vacant halls, where echoes whisper low,\nA ...   

                              Mistral 7B INSTRUCT Q4  \
0  Upon the field of battle, where chaos reigns,\...   
1  In the vast expanse of the cosmos, where stars...   
2  In the empty home, where laughter once was hea...   

                                       Gemma 2 2B Q4  \
0  Upon the field of strife, where blood doth sta...   
1  Upon the velvet canvas of the night,\nA millio...   
2  The dust motes dance in sunbeams, pale and thi...   

                                QWEN2 7B Instruct Q4  \
0  Amidst the clash of stee

In [5]:
print("\nColumn Names:")
print(data.columns)


Column Names:
Index(['Prompt/Model', 'Llama 3.2 3B Instruct Q8', 'Mistral 7B INSTRUCT Q4',
       'Gemma 2 2B Q4', 'QWEN2 7B Instruct Q4', 'Openchat-3.5 7B',
       'Code Ninja 7B Q4'],
      dtype='object')


In [6]:
print("\nDataset Info:")
data.info()


Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 7 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Prompt/Model              3 non-null      object
 1   Llama 3.2 3B Instruct Q8  3 non-null      object
 2   Mistral 7B INSTRUCT Q4    3 non-null      object
 3   Gemma 2 2B Q4             3 non-null      object
 4   QWEN2 7B Instruct Q4      3 non-null      object
 5   Openchat-3.5 7B           3 non-null      object
 6   Code Ninja 7B Q4          3 non-null      object
dtypes: object(7)
memory usage: 300.0+ bytes


### Reference Poems

A list of reference poems is provided to evaluate the generated poem's quality.

The reference for the poem in prompt 1 is taken from the website AllPoetry[https://allpoetry.com/poems/about/Courage]. The poem is titled The Brave Ones and is authored by Guilty.

In [7]:
referenced_poem_prompt1 = (
    "The brave ones are the living"
    "grabbing life with both hands"
    "reckless adventures flying"
    "with not a care where they land"
    
    "The meek ones merely exist"
    "going unnoticed is their plan"
    "perhaps they will perish"
    "before they need to take a stand"
    
    "The brave ones believe"
    "in more than what they see"
    "the great unknown mystifies"
    "and life is yet to be revealed"
    
    "The meek ones pity their lot"
    "whatever they have is what they have got"
    "there is nothing grand"
    "in eating out of another’s hands"
    
    "Be bold, be brave and take a chance"
    "a glorious failure is better than a lofty dream"
    "never aspired for"
)

The reference for the poem in prompt 2 is taken from the website [https://poemsplease.com/12-poems-reflecting-the-wonder-of-space-exploration/]. The poem is titled Boundless Skies. 

In [8]:
referenced_poem_prompt2 = (
    "In the stillness of night, where silence reigns,"
    "A canvas spreads, inked with ethereal gains."
    "Stars like diamonds twinkle, shimmering light,"
    "Infinite stories held, ready for flight."
    "Comets trace paths, fiery trails that gleam,"
    "While galaxies swirl in the vastest dream."
    "Cosmic whispers call, beckoning the brave,"
    "To wander through wonders, the universe to save."
)

The reference for the poem in prompt 3 is taken from the website [https://statics.teams.cdn.office.net/evergreen-assets/safelinks/1/atp-safelinks.html]. The poem is titled The Empty House and is authored by Walter de La Mare.

In [9]:
referenced_poem_prompt3 = (
    "Whenever I walk to Suffern along the Erie track"
    "I go by a poor old farmhouse with its shingles broken and black."
    "I suppose I've passed it a hundred times, but I always stop for a minute"
    "And look at the house, the tragic house, the house with nobody in it."
    
    "I never have seen a haunted house, but I hear there are such things;"
    "That they hold the talk of spirits, their mirth and sorrowings."
    "I know this house isn't haunted, and I wish it were, I do;"
    "For it wouldn't be so lonely if it had a ghost or two."
    
    "This house on the road to Suffern needs a dozen panes of glass,"
    "And somebody ought to weed the walk and take a scythe to the grass."
    "It needs new paint and shingles, and the vines should be trimmed and tied;"
    "But what it needs the most of all is some people living inside."
    
    "If I had a lot of money and all my debts were paid"
    "I'd put a gang of men to work with brush and saw and spade."
    "I'd buy that place and fix it up the way it used to be"
    "And I'd find some people who wanted a home and give it to them free."
    
    "Now, a new house standing empty, with staring window and door,"
    "Looks idle, perhaps, and foolish, like a hat on its block in the store."
    "But there's nothing mournful about it; it cannot be sad and lone"
    "For the lack of something within it that it has never known."
    
    "But a house that has done what a house should do,"
    "a house that has sheltered life,"
    "That has put its loving wooden arms around a man and his wife,"
    "A house that has echoed a baby's laugh and held up his stumbling feet,"
    "Is the saddest sight, when it's left alone, that ever your eyes could meet."
    
    "So whenever I go to Suffern along the Erie track"
    "I never go by the empty house without stopping and looking back,"
    "Yet it hurts me to look at the crumbling roof and the shutters fallen apart,"
    "For I can't help thinking the poor old house is a house with a broken heart."
)


In [10]:
reference_poems = [referenced_poem_prompt1, referenced_poem_prompt2, referenced_poem_prompt3]

In [11]:
data["Reference Poem"] = reference_poems

In [12]:
data.head()

Prompt/Model  \
0  Write a Shakespearean sonnet \nabout courage s...   
1  Write a Shakespearean sonnet \nabout wonder se...   
2  Write a Shakespearean sonnet \nabout loss set ...   

                            Llama 3.2 3B Instruct Q8  \
0  Fair battlefield, where valorous hearts do lie...   
1  Fairest cosmos, thou dost stretch thy might,\n...   
2  In vacant halls, where echoes whisper low,\nA ...   

                              Mistral 7B INSTRUCT Q4  \
0  Upon the field of battle, where chaos reigns,\...   
1  In the vast expanse of the cosmos, where stars...   
2  In the empty home, where laughter once was hea...   

                                       Gemma 2 2B Q4  \
0  Upon the field of strife, where blood doth sta...   
1  Upon the velvet canvas of the night,\nA millio...   
2  The dust motes dance in sunbeams, pale and thi...   

                                QWEN2 7B Instruct Q4  \
0  Amidst the clash of steel and blood, I stand u...   
1  In the vast expanse of cosmic air, where stars...   
2  In silent halls where shadows loom, \nThe echo...   

                                     Openchat-3.5 7B  \
0  Upon the field of strife, where metal meets,\n...   
1  Upon the canvas of the boundless sky,\nWhere d...   
2  Upon these hallowed walls, where echoes linger...   

                                    Code Ninja 7B Q4  \
0  Upon the battlefield where fierce armies clash...   
1  When first upon this boundless void I gazed,\n...   
2  When shadows eclipse the sun's bright ray,\nAn...   

                                      Reference Poem  
0  The brave ones are the livinggrabbing life wit...  
1  In the stillness of night, where silence reign...  
2  Whenever I walk to Suffern along the Erie trac...

## Fluency

Evaluating fluency is important in poetry generation to ensure that the generated poems not only follow linguistic rules but also convey meaning smoothly. Several metrics can be used to assess the fluency of generated poetry:
*   BLEU (Bilingual Evaluation Understudy):
Definition  : BLEU is a metric commonly used in machine translation. It measures the precision of n-grams (contiguous sequences of n words) in the generated text against a reference text.
Use         : While it’s more commonly applied to translation tasks, it can be adapted to poetry evaluation by comparing n-grams between generated poems and reference poems.

*   METEOR (Metric for Evaluation of Translation with Explicit ORdering):
Definition  : METEOR is designed to address some of the shortcomings of BLEU, particularly by considering synonyms, stemming, and word order.
Use         : It evaluates the semantic and syntactic similarity between generated and reference text, which can be helpful for poetry generation.

*   Perplexity:
Definition  : Perplexity measures how well a model predicts the next word in a sequence. Lower perplexity indicates that the model is better at predicting text and is more "certain" in its predictions.
Use         : This can give an idea of how fluent or coherent the generated poem is, although it doesn't always align with creative quality.

*   Entropy:  
Definition  : Entropy quantifies the average uncertainty or information content in the text. It measures the diversity or randomness in token distributions, such as characters or words, in the generated output. Higher entropy indicates more diverse text, while lower entropy suggests repetitive or predictable output.
Use         : Entropy is valuable for evaluating poetry generation as it reflects the balance between creativity and structure. For poetry, moderate entropy often indicates well-crafted and varied text with coherent patterns, while extreme entropy values can signal overly repetitive or chaotic outputs. This metric helps compare models’ ability to produce diverse and engaging poems.  

In [13]:
# Evaluation Functions
def bleu_score(reference, candidate):
    return round(sentence_bleu([reference.split()], candidate.split()), 5)

def meteor_score_func(reference, candidate):
    # Tokenize both the reference and candidate
    reference_tokens = reference.split()  
    candidate_tokens = candidate.split()  
    return round(meteor_score([reference_tokens], candidate_tokens), 5)

def calculate_perplexity(text):
    inputs = tokenizer(text, return_tensors="pt")
    with torch.no_grad():
        outputs = model(**inputs, labels=inputs["input_ids"])
        loss = outputs.loss
        return round(torch.exp(loss).item(), 5)

def calculate_entropy(text):
    tokens = list(text)  
    
    token_counts = Counter(tokens)
    total_tokens = len(tokens)
    
    probabilities = [count / total_tokens for count in token_counts.values()]
    
    entropy = -sum(p * math.log2(p) for p in probabilities)
    return round(entropy, 5)

In [14]:
# Specify the model column to evaluate
model_name_to_evaluate = "Mistral 7B INSTRUCT Q4"  

In [16]:
# Evaluate Scores
results = []
for index, row in data.iterrows():
    prompt = row["Prompt/Model"]
    reference_poem = row["Reference Poem"]
    generated_poem = row[model_name_to_evaluate]
    
    # Calculate scores
    bleu = bleu_score(reference_poem, generated_poem)
    meteor = meteor_score_func(reference_poem, generated_poem)
    perplexity = calculate_perplexity(generated_poem)
    entropy = calculate_entropy(generated_poem)
    
    # Append to results
    results.append({
        "Prompt": prompt,
        "BLEU": bleu,
        "METEOR": meteor,
        "Perplexity": perplexity,
        "Entropy": entropy  # Fixed key capitalization for clarity
    })

# Convert results to DataFrame for better visualization
results_df = pd.DataFrame(results)

# Calculate average entropy
average_entropy = results_df["Entropy"].mean()
average_perplexity = results_df["Perplexity"].mean()

# Add model name to the title
evaluation_title = f"Evaluation Results for Model: {model_name_to_evaluate}"

# Display Results
print(evaluation_title)
print(results_df)

# Display the average entropy
print(f"\nAverage Entropy: {average_entropy:.4f}")
print(f"\nAverage Perplexity: {average_perplexity:.4f}")


Evaluation Results for Model: Mistral 7B INSTRUCT Q4
                                              Prompt  BLEU   METEOR  \
0  Write a Shakespearean sonnet \nabout courage s...   0.0  0.06486   
1  Write a Shakespearean sonnet \nabout wonder se...   0.0  0.19382   
2  Write a Shakespearean sonnet \nabout loss set ...   0.0  0.08349   

   Perplexity  Entropy  
0     2.95177  4.39219  
1     3.01616  4.38685  
2     2.66690  4.31930  

Average Entropy: 4.3661

Average Perplexity: 2.8783



1. **BLEU**: This metric is 0.0 for all three poems, indicating that there is no repetition in the generated text. It suggests that the model has created highly unique text for each prompt without producing identical outputs.

2. **METEOR**:METEOR scores measure the similarity between the generated text and reference text. A higher score suggests better alignment with reference human-written text. Poem 2 has a higher METEOR score(0.19382), implying it is more similar to the reference text than the others.

3. **Perplexity**:Perplexity measures how well the model predicts the next word in a sequence. A lower perplexity indicates better performance. Poem 3 has the lowest perplexity, suggesting it is the easiest for the model to predict and thus has higher coherence or fluency. The average perplexity value of 2.8783 for the Mistral model indicates a relatively low level of uncertainty in its predictions, suggesting that the model generates more fluent and coherent text with fewer surprises.

4. **Entropy**: Entropy reflects the uncertainty or randomness of the generated text. A higher value suggests more unpredictability. Poem 1 has the highest entropy, suggesting it is the most varied or unpredictable text among the three, while Poem 3 has slightly lower entropy, implying more predictable output.The average entropy value of 4.3661 for the  model indicates a moderate level of unpredictability in its predictions, suggesting that the model's generated text has some variability or uncertainty.

### Diversity and Variety

In the context of poem generation, diversity and variety refer to the richness, uniqueness, and creativity of the generated text, which are critical for producing compelling, engaging, and original poetry.

* Self-BLEU : Evaluates how much overlap exists between different samples generated by the same model.
* Distinct n-grams: Measures the percentage of distinct n-grams in the generated text compared to the total n-grams. Higher distinctness means more diversity.

In [17]:
import pandas as pd
import collections
from nltk.translate.bleu_score import sentence_bleu
from nltk.util import ngrams

In [18]:
# Self-BLEU Score
def self_bleu(generated_texts, n=1):
    all_ngrams = []
    for text in generated_texts:
        if not isinstance(text, str):
            text = str(text)  # Ensure text is a string
        tokens = text.split()  # Tokenize text
        ngrams_list = list(ngrams(tokens, n))  # Create n-grams
        all_ngrams.append(collections.Counter(ngrams_list))
    
    # Compute Self-BLEU score
    score = 0
    for i, counter in enumerate(all_ngrams):
        others = all_ngrams[:i] + all_ngrams[i+1:]  # All other generated poems except the current one
        union_ngrams = collections.Counter()
        for other in others:
            union_ngrams.update(other)
        
        # Avoid division by zero
        if sum(union_ngrams.values()) > 0:
            overlap = sum(min(count, union_ngrams[gram]) for gram, count in counter.items())
            score += overlap / sum(union_ngrams.values())
    
    return score / len(generated_texts) if len(generated_texts) > 1 else 0

# Distinct-N Score
def distinct_n(generated_texts, n=2):
    ngrams_set = set()
    total_ngrams = 0
    
    for text in generated_texts:
        if not isinstance(text, str):
            text = str(text)  # Ensure text is a string
        tokens = text.split()  # Tokenize text
        ngrams_list = list(ngrams(tokens, n))  # Create n-grams
        ngrams_set.update(ngrams_list)  # Add n-grams to the set
        total_ngrams += len(ngrams_list)
    
    return len(ngrams_set) / total_ngrams if total_ngrams > 0 else 0

In [19]:
# Function to evaluate Self-BLEU and Distinct-N for a model
def evaluate_model_metrics(data, model_name_to_evaluate):
    results = []

    # Collect all poems generated by the same model across different prompts
    generated_poems = data[model_name_to_evaluate].tolist()

    # Calculate Self-BLEU and Distinct-N for the poems generated by the model
    self_bleu_score = self_bleu(generated_poems, n=1)  # For unigrams (Self-BLEU n=1)
    distinct_2_score = distinct_n(generated_poems, n=2)  # For distinct-2 (bigrams)
    
    for index, row in data.iterrows():
        prompt = row["Prompt/Model"]
        
        results.append({
            "Model": model_name_to_evaluate,
            "Prompt": prompt,
            "Self-BLEU (n=1)": round(self_bleu_score, 5),
            "Distinct-2": round(distinct_2_score, 5)
        })

    results_df = pd.DataFrame(results)
    
    evaluation_title = f"Evaluation Results for Model: {model_name_to_evaluate}"
    
    print(evaluation_title)
    print(results_df)

# Example usage
model_name_to_evaluate = "Mistral 7B INSTRUCT Q4"  
evaluate_model_metrics(data, model_name_to_evaluate)

Evaluation Results for Model: Mistral 7B INSTRUCT Q4
                    Model                                             Prompt  \
0  Mistral 7B INSTRUCT Q4  Write a Shakespearean sonnet \nabout courage s...   
1  Mistral 7B INSTRUCT Q4  Write a Shakespearean sonnet \nabout wonder se...   
2  Mistral 7B INSTRUCT Q4  Write a Shakespearean sonnet \nabout loss set ...   

   Self-BLEU (n=1)  Distinct-2  
0          0.20741     0.74734  
1          0.20741     0.74734  
2          0.20741     0.74734  



1. **Self-BLEU**: 
   - The Self-BLEU score for all three prompts is **0.20741**. This metric measures the uniqueness of the generated text by comparing it to other parts of the generated text. A score of 0.20741 suggests moderate repetition, meaning the model does not excessively repeat itself but may still reuse some structures or phrases across different outputs.

2. **Distinct-2**: 
   - The **Distinct-2** score is **0.74734** for all three prompts. This metric evaluates the lexical diversity of the generated text by considering the proportion of unique 2-grams (pairs of consecutive words) in the output. A higher score indicates more varied output. In this case, the score of 0.74734 shows that the model generates fairly diverse text with a good amount of unique word pairs, contributing to more varied and less repetitive outputs.

### Cleaning the sonnet and storing as a list

In [20]:
import pandas as pd
import string

# Function to clean and split poems
def process_poem(poem):
    if isinstance(poem, str):  # Ensure the input is a string
        lines = [
            line.translate(str.maketrans('', '', string.punctuation))
            for line in poem.strip().split('\n') if line.strip()
        ]
        return poem, lines
    else:
        return None, []  # Handle NaN or non-string values


# Filter the data for the selected model
selected_poems = data[model_name_to_evaluate].head(3)  # Select the first 3 poems for the model

# Process the selected poems and assign lines to variables
poem1, lines1 = process_poem(selected_poems.iloc[0])
poem2, lines2 = process_poem(selected_poems.iloc[1])
poem3, lines3 = process_poem(selected_poems.iloc[2])

# Print the processed lines for poem1
print("Poem 1:")
print(f"  Original Poem: {poem1}")
print(f"  Processed Lines: {lines1}")


Poem 1:
  Original Poem: Upon the field of battle, where chaos reigns,
Steps forth a figure, with courage unstained.
A warrior, a hero, a beacon of might,
Against the darkness, he stands in the fight.

His eyes, they glimmer, with a flame so bright,
A symbol of hope, a beacon of light.
His spirit unbroken, his heart unbowed,
He marches on, with courage unstowed.

The bullets fly, the swords clash, the drums beat,
But he remains undaunted, his courage complete.
He fights for freedom, for justice, for right,
A champion of hope, a symbol of light.

In the midst of the chaos, he stands tall,
His courage, his strength, his unyielding call.
A symbol of resilience, of strength untold,
A hero, a warrior, a true guardian of gold.

  Processed Lines: ['Upon the field of battle where chaos reigns', 'Steps forth a figure with courage unstained', 'A warrior a hero a beacon of might', 'Against the darkness he stands in the fight', 'His eyes they glimmer with a flame so bright', 'A symbol of hope a b

### Poetic Structure:

A Shakespearean Sonnet is a 14 line poem which follows a consitent rhyme scheme pattern of ABAB CDCD EFEF GG and each line is 10 syllables long written in iambic pentameter.

#### Finding the Rhyming Scheme

In [21]:
import pronouncing

poem_list = [lines1, lines2, lines3]

def get_rhyming_scheme(lines):
    # Extract the last word of each line
    last_words = [line.split()[-1] for line in lines]
    rhyme_dict = {}
    scheme = []
    current_letter = 'A'

    for word in last_words:
        rhymes = pronouncing.rhymes(word)
        # Check if the word rhymes with any previously seen words
        for key, letter in rhyme_dict.items():
            if word in rhymes or key in rhymes:
                scheme.append(letter)
                break
        else:
            # Assign a new letter if no rhyming word is found
            rhyme_dict[word] = current_letter
            scheme.append(current_letter)
            current_letter = chr(ord(current_letter) + 1)

    rhyme_scheme = ''.join(scheme)
    return '\n'.join([rhyme_scheme[i:i+4] for i in range(0, len(rhyme_scheme), 4)])


# Process each poem and print rhyming schemes
for i, poem in enumerate(poem_list, 1):
    rhyme_scheme = get_rhyming_scheme(poem)
    print(f"Rhyming Scheme for Poem {i}: \n{rhyme_scheme}")


Rhyming Scheme for Poem 1: 
ABCC
CCDE
FFCC
GGHH
Rhyming Scheme for Poem 2: 
AABC
DEFF
AAGG
HH
Rhyming Scheme for Poem 3: 
ABCC
DDEE
FGHH
IJKK


When we compare the rhyming schemes of these poems to a Shakespearean sonnet, we can see clear differences:

**Poem 1**:Rhyming Scheme: ABCC CCDE FFCC GGHH
It deviates from the Shakespearean sonnet by introducing repeated rhymes (CC and FFCC) within its structure, which does not follow the alternating pattern of a Shakespearean sonnet.
The poem ends with GGHH, where the last two lines introduce an additional rhyme (HH), deviating from the traditional GG couplet.

**Poem 2**:Rhyming Scheme: AABC DEFF AAGG HH
It introduces a non-standard rhyme structure with AABC, which is not typical for a Shakespearean sonnet.
It ends with HH, but the rhymes used in the rest of the poem (AABC, DEFF, AAGG) differ from the consistent alternating Stanzas of a Shakespearean sonnet.

**Poem 3**:Rhyming Scheme: ABCC DDEE FGHH IJKK
It shows a highly irregular rhyme pattern, with no alternating rhymes and new rhymes introduced in each section.
It ends with IJKK, which does not conform to the Shakespearean sonnet's GG couplet, and contains no clear alternating Stanzas.

#### Finding the Syllable Counts

In [22]:
import nltk
from nltk.corpus import cmudict
from collections import defaultdict
import re

nltk.download("cmudict")
cmu_dict = cmudict.dict()

def count_syllables(word):
    """
    Count syllables for a given word using the CMU Pronouncing Dictionary.
    """
    word = word.lower()
    if word in cmu_dict:
        return min([len([y for y in x if y[-1].isdigit()]) for x in cmu_dict[word]])
    return 1  

def analyze_syllable_counts(poem):
    """
    Analyze the syllable count for each line in the poem.
    """
    syllable_counts = [sum(count_syllables(word) for word in re.findall(r'\w+', line)) for line in poem]
    return syllable_counts

all_syllable_counts = [analyze_syllable_counts(poem) for poem in poem_list]

for idx, syllable_count in enumerate(all_syllable_counts):
    print(f"Poem {idx + 1} syllable counts: {syllable_count}")

[nltk_data] Downloading package cmudict to
[nltk_data]     C:\Users\pooji\AppData\Roaming\nltk_data...
[nltk_data]   Package cmudict is already up-to-date!


Poem 1 syllable counts: [11, 10, 11, 10, 10, 10, 10, 8, 10, 12, 10, 11, 10, 10, 11, 13]
Poem 2 syllable counts: [14, 14, 10, 9, 13, 12, 11, 10, 10, 10, 11, 11, 13, 15]
Poem 3 syllable counts: [11, 8, 10, 14, 10, 11, 11, 11, 12, 10, 9, 10, 11, 8, 10, 14]


The syllable counts across the three poems show significant variation, deviating from the traditional iambic pentameter structure of a Shakespearean sonnet (10 syllables per line).

- **Poem 1**: Has a mix of 8-13 syllables per line, indicating some irregularity in rhythm.
- **Poem 2**: Shows wider fluctuations (9-15 syllables), suggesting free-form and less structured rhythm.
- **Poem 3**: Varies between 8-14 syllables, with some lines being notably shorter or longer, resulting in an irregular meter.